kubelet/datadog_checks/kubelet/kubelet.py

```diff
@@ -291,10 +300,13 @@ def _report_pods_running(self, pods, instance_tags):
                 tags += instance_tags
                 hash_tags = tuple(sorted(tags))
                 containers_tag_counter[hash_tags] += 1
+                containers_restarts_tag_counter[hash_tags] += container.get('restartCount', 0)
```

Comment:
That was definitely my goal (per pod/container high cardinality), but I don't really understand how to figure out which tags are available in which contexts from `get_tags`.  I was basing this on observing that the container spec metrics which get tags via https://github.com/DataDog/integrations-core/blob/407294e86069d8b70745ad37a1b67faff1d35c92/kubelet/datadog_checks/kubelet/kubelet.py#L331 seem to have tags for kube_pod_name, kube_container_name, image_id, etc.

Separate from whether the tags are correct, moving this to `_report_container_state_metrics` could be better from a code organization standpoint.

Reply:
In this loop, container tags are retrieved in low cardinality (`tags = get_tags(container_id, False)`). Low cardinality excludes tags like `container_name`, `container_id`, `pod_name` that churns when a pod would be re-created with the same spec. You would basically get the same tags as in AutoDiscovery (kube_deployment, kube_container_name, kube_namespace, docker_image...).

What we want here is to be able to identify what exact pod might be restarting, instead of just the deployment, this is why I suggested using high cardinality for these. In high cardinality, there is no need to go through a counter, as we will have one timeserie per container, the logic would be closer to what we have in [`_report_container_spec_metrics`](https://github.com/DataDog/integrations-core/blob/cb931199d2fc9982339d789ed7403361ab605cc0/kubelet/datadog_checks/kubelet/kubelet.py#L305) (iterate over pods, then containers, directly submit gauges for each)

If you want to dig into what tag is what, you can have a look at our tag extractors:

  - kubelet: https://github.com/DataDog/datadog-agent/blob/master/pkg/tagger/collectors/kubelet_extract.go
  - docker: https://github.com/DataDog/datadog-agent/blob/master/pkg/tagger/collectors/docker_extract.go

You'll see tags added with either `AddLow` or `AddHigh`.

Reply:
Thank you for the detailed explanation, I completely missed the high/low distinction.  I've moved the restarts metrics to the state metrics.